File: js/src/rml-check.mjs
#!/usr/bin/env node
// rml-check — independent proof-replay checker (issue #36).
//
// CLI front-end for the `check.mjs` module. Reads a program file and a
// proof file, validates that every derivation in the proof file
// corresponds to a query in the program, and prints
// `OK: N derivations replayed.` on success.

import fs from 'node:fs';
import { checkProgram, isOk } from './check.mjs';

function main(argv) {
  if (argv.length !== 4) {
    process.stderr.write('Usage: rml-check <program.lino> <proofs.lino>\n');
    return 2;
  }
  const programPath = argv[2];
  const proofsPath = argv[3];

  let program;
  let proofs;
  try {
    program = fs.readFileSync(programPath, 'utf8');
  } catch (e) {
    process.stderr.write(`Error reading ${programPath}: ${e.message}\n`);
    return 1;
  }
  try {
    proofs = fs.readFileSync(proofsPath, 'utf8');
  } catch (e) {
    process.stderr.write(`Error reading ${proofsPath}: ${e.message}\n`);
    return 1;
  }

  const result = checkProgram(program, proofs);
  if (isOk(result)) {
    process.stdout.write(`OK: ${result.ok.length} derivations replayed.\n`);
    return 0;
  }
  for (const err of result.errors) {
    const path = err.path.length === 0 ? '<root>' : err.path.join(' / ');
    process.stderr.write(`FAIL [${path}]: ${err.message}\n`);
  }
  return 1;
}

process.exit(main(process.argv));

File: rust/src/bin/rml-check.rs
// rml-check — independent proof-replay checker (issue #36).
//
// CLI front-end for the `rml::check` module. Reads a program file and a
// proof file, validates that every derivation in the proof file
// corresponds to a query in the program, and prints `OK: N derivations
// replayed.` on success.

use rml::check::check_program;
use std::env;
use std::fs;
use std::process::ExitCode;

fn main() -> ExitCode {
    let args: Vec<String> = env::args().collect();
    if args.len() != 3 {
        eprintln!("Usage: rml-check <program.lino> <proofs.lino>");
        return ExitCode::from(2);
    }
    let program_path = &args[1];
    let proofs_path = &args[2];
    let program = match fs::read_to_string(program_path) {
        Ok(t) => t,
        Err(e) => {
            eprintln!("Error reading {}: {}", program_path, e);
            return ExitCode::from(1);
        }
    };
    let proofs = match fs::read_to_string(proofs_path) {
        Ok(t) => t,
        Err(e) => {
            eprintln!("Error reading {}: {}", proofs_path, e);
            return ExitCode::from(1);
        }
    };
    let result = check_program(&program, &proofs);
    if result.is_ok() {
        println!("OK: {} derivations replayed.", result.ok.len());
        ExitCode::SUCCESS
    } else {
        for err in &result.errors {
            let path = if err.path.is_empty() {
                String::from("<root>")
            } else {
                err.path.join(" / ")
            };
            eprintln!("FAIL [{}]: {}", path, err.message);
        }
        ExitCode::from(1)
    }
}
